src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs

```diff
@@ -198,12 +191,10 @@ public virtual void OnSocketClosed()
 
         private void ApplyConnectionFilter()
```

Comment:
There was a very obvious deadlock here too: if `_connectionState != ConnectionState.CreatingFrame` we would call `ConnectionControl.End`, which would try to lock `_stateLock`.

I'm not sure if the lock is required only for synchronizing `_connectionState` or if everything else should be inside the lock, so to be safe I moved the `lock` statement to the `if` body after atomically updating `_connectionState`.